Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set here, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted


We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

* Ask HN: How to improve my personal website?
* Ask HN: Am I the only one outraged by Twitter shutting down share counts?
* Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

* Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
* Show HN: Something pointless I made
* Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
Let's start by importing the libraries we need and reading the data set into a list of lists.

#### INSTRUCTIONS
1. Start by adding a title and writing a paragraph in a markdown cell introducing the project and the data set. The title and the introduction are tentative at this point, so don't spend too much time here — you can come back at the end of your work to refine them.
2. Read the hacker_news.csv file in as a list of lists.
    * Assign the result to the variable hn.
3. Display the first five rows of hn.

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
hn [:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

Above, we read our data into a list of lists called hn. When you displayed the first five rows of hn, you should've seen the following data:

[
    ['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], 
    ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], 
    ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], 
    ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], 
    ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']
]

Notice that the first list in the inner lists contains the column headers, and the lists after contain the data for one row. In order to analyze our data, we need to first remove the row containing the column headers. Let's remove that first row next.


### INSTRUCTIONS

1. Extract the first row of data, and assign it to the variable headers.
2. Remove the first row from hn.
3. Display headers.
4. Display the first five rows of hn to verify that you removed the header row properly.


In [2]:
headers = hn[:1]
hn = hn[1:]
print ("Display headers\n")
print (headers)
print ("\n display first five rows of hn\n")
print (hn[:5])

Display headers

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]

 display first five rows of hn

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2

Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

To find the posts that begin with either Ask HN or Show HN, we'll use the string method startswith. Given a string object, say, string1, we can check if starts with, say, dq by inspecting the output of the object string1.startswith('dq'). If string1 starts with dq, it will return True, otherwise it will return False.


In [3]:
print('dataquest'.startswith('Data'))

False


In [4]:
print('dataquest'.startswith('data'))

True


In the example above, the first print call gives us False because dataquest does not start with Data. The second print call prints True because dataquest does start with data. Capitalization matters.

If we wish to control for case, we can use the lower method which returns a lowercase version of the starting string. Here's an example:


In [5]:
print('Dataquest'.lower())

dataquest


Let's use these methods to separate posts beginning with Ask HN and Show HN (and case variations) into two different lists next.

#### Instructions


1. Create three empty lists called ask_posts, show_posts, and other_posts.
2. Loop through each row in hn.
 * Assign the title in each row to a variable named title.
      *Because the title column is the second column, you'll need to get the element at index 1 in each row.
3. Implement the following steps:
 * If the lowercase version of title starts with ask hn, append the row to ask_posts.
 * Else if the lowercase version of title starts with show hn, append the row to show_posts.
 * Else append to other_posts.
4. Check the number of posts in ask_posts, show_posts, and other_posts.

In [6]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
        
print ("Number of posts in ask_posts :")
print (len(ask_posts))
print ("\nNumber of posts in show_posts: ")
print (len(show_posts))
print ("\nNumber of posts in other_posts:")
print (len(other_posts))


        

Number of posts in ask_posts :
1744

Number of posts in show_posts: 
1162

Number of posts in other_posts:
17194


In the last screen, we separated the "ask posts" and the "show posts" into two list of lists named ask_posts and show_posts. Below are the first five rows in the ask_posts list of lists:

In [7]:
print ("First five rows of ask_posts")
print (ask_posts[:5])
print ("\nFirst five rows of show_posts")
print (show_posts[:5])


First five rows of ask_posts
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]

First five rows of show_posts
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming pl

Next, let's determine if ask posts or show posts receive more comments on average.

### INSTRUCTIONS: 


1. Find the total number of comments in ask posts and assign it to total_ask_comments.
   * Set total_ask_comments to 0.
2. Use a for loop to iterate over the ask posts.
   * Because the num_comments column is the fifth column in ask_posts, you'll need to get the element at index 4 in each row.
      *  You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
      *  Add this value to total_ask_comments.
3. Compute the average number of comments on ask posts and assign it to avg_ask_comments.
4. Print avg_ask_comments.
5. Find the total number of comments in show posts and assign it to total_show_comments.
   * Set total_show_comments to 0.
6. Use a for loop to iterate over the ask posts.
    * Because the num_comments column is the fifth column in show_posts, you'll need to get the element at index 4 in each row.
       * You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
       * Add this value to total_show_comments.
7. Compute the average number of comments on show posts and assign it to avg_show_comments.
8. Print avg_show_comments.
9. Do show posts or ask posts receive more comments on average? Write a markdown cell explaining your findings.


In [8]:
total_ask_comments = 0
for each in ask_posts:
    num_comments = int(each[4])
    total_ask_comments += num_comments
    
print ("Total_ask_comments: ", total_ask_comments )
avg_ask_comments = total_ask_comments / len(ask_posts)
print ("Average number of comments on ask posts: ", avg_ask_comments)



Total_ask_comments:  24483
Average number of comments on ask posts:  14.038417431192661


In [9]:
total_show_comments = 0
for each in show_posts:
    num_comments = int(each[4])
    total_show_comments += num_comments

print ("Total_show_comments: ", total_show_comments )
avg_show_comments = total_show_comments / len(show_posts)
print ("Average number of comments on show posts: ", avg_show_comments)

Total_show_comments:  11988
Average number of comments on show posts:  10.31669535283993


#### INFERENCE
On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

In the last screen, you should've determined that, on average, ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

In this screen, we'll tackle the first step — calculating the amount of ask posts and comments by hour created. We'll use the datetime module to work with the data in the created_at column.

Recall that we can use the datetime.strptime() constructor to parse dates stored as strings and return datetime objects:

date_1_str = "December 24, 1984"
date_1_dt = dt.datetime.strptime(date_1_str, "%B %d, %Y")

Note in the example above that %B represents the full name of the month, %d represents a two digit day, and %Y represents the four digit year. If you need to parse a string in a different format, you'll have to modify these codes.

Let's use this technique to calculate the amount of ask posts created per hour, along with the total amount of comments.

#### INSTRUCTIONS:

1. Import the datetime module as dt.
2. Create an empty list and assign it to result_list. This will be a list of lists.
3. Iterate over ask_posts and append to result_list a list with two elements:
    * The first element shall be the column created_at.
        * Because the created_at column is the seventh column in ask_posts, you'll need to get the element at index 6 in each row.
    * The second element shall be the number of comments of the post.
        * You'll also need to convert the value to an integer
4. Create two empty dictionaries called counts_by_hour and comments_by_hour.
5. Loop through each row of result_list.
6. Extract the hour from the date, which is the first element of the row.
7. Use the datetime.strptime() method to parse the date and create a datetime object.
8. Use the string we want to parse as the first argument and a string that specifies the format as the second argument.
    * Use the datetime.strftime() method to select just the hour from the datetime object.
    * If the hour isn't a key in counts_by_hour:
        * Create the key in counts_by_hour and set it equal to 1.
        * Create the key in comments_by_hour and set it equal to the comment number.
    * If the hour is already a key in counts_by_hour:
        * Increment the value in counts_by_hour by 1.
        * Increment the value in comments_by_hour by the comment number.


In [10]:
import datetime as dt


In [11]:
result_list = []
for each in ask_posts:
    result_list.append([each[6],int(each[4])])

print (result_list[:5])
    
    

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [12]:
counts_by_hour = {}
comments_by_hour = {}
#If you see the list result_list the date and time format is as below
# eg: '8/16/2016 9:55'
# HENCE initialize a date_format with the above format
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    num_of_comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += num_of_comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = num_of_comment
        counts_by_hour[time] = 1
        
print (comments_by_hour)
print (counts_by_hour)

{'11': 641, '23': 543, '05': 464, '18': 1439, '13': 1253, '04': 337, '21': 1745, '02': 1381, '03': 421, '01': 683, '09': 251, '08': 492, '14': 1416, '00': 447, '15': 4477, '06': 397, '10': 793, '16': 1814, '17': 1146, '12': 687, '07': 267, '20': 1722, '22': 479, '19': 1188}
{'11': 58, '23': 68, '05': 46, '18': 109, '13': 85, '04': 47, '21': 109, '02': 58, '03': 54, '01': 60, '09': 45, '08': 48, '14': 107, '00': 55, '15': 116, '06': 44, '10': 59, '16': 108, '17': 100, '12': 73, '07': 34, '20': 80, '22': 71, '19': 110}


In the last screen, we created two dictionaries:

* counts_by_hour: contains the number of ask posts created during each hour of the day.
* comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.

Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

To illustrate the technique, let's work with the following dictionary:

In [13]:
sample_dict = {
                'apple': 2, 
                'banana': 4, 
                'orange': 6
               }


Suppose we wanted to multiply each of the values by ten and return the results as a list of lists. We can use the following code:

In [14]:
fruits = []

for fruit in sample_dict:
    fruits.append([fruit, 10*sample_dict[fruit]])

print ("RESULTS: \n")
print (fruits)

RESULTS: 

[['apple', 20], ['orange', 60], ['banana', 40]]


In the example above, we:

* Initialized an empty list (of lists) and assigned it to fruits.
* Iterated over the keys of sample_dict and appended to fruits a list whose:
    * First element is the key from sample_dict.
    * Second element is the value corresponding to that key multiplied by ten.

Let's use this format to create a list of lists containing the hours during which posts were created and the average number of comments those posts received.

#### INSTRUCTIONS:

1. Use the example above to calculate the average number of comments per post for posts created during each hour of the day.
2. The result should be a list of lists in which the first element is the hour and the second element is the average number of comments per post. Assign the result to a variable named avg_by_hour. Display the results.

In [15]:
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
    
    
avg_by_hour

[['11', 11.051724137931034],
 ['23', 7.985294117647059],
 ['05', 10.08695652173913],
 ['18', 13.20183486238532],
 ['13', 14.741176470588234],
 ['04', 7.170212765957447],
 ['21', 16.009174311926607],
 ['02', 23.810344827586206],
 ['03', 7.796296296296297],
 ['01', 11.383333333333333],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['14', 13.233644859813085],
 ['00', 8.127272727272727],
 ['15', 38.5948275862069],
 ['06', 9.022727272727273],
 ['10', 13.440677966101696],
 ['16', 16.796296296296298],
 ['17', 11.46],
 ['12', 9.41095890410959],
 ['07', 7.852941176470588],
 ['20', 21.525],
 ['22', 6.746478873239437],
 ['19', 10.8]]

In the last screen, we calculated the average number of comments for posts created during each hour of the day, and stored the results in a list of lists named avg_by_hour as above. 

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

#### INSTRUCTIONS

1. Create a list that equals avg_by_hour with swapped columns.
    * Create an empty list and assign it to swap_avg_by_hour.
    * Iterate over the rows of avg_by_hour and append to swap_avg_by_hour a list whose first element is the second element of the row, and whose second element is the first element of the row.
2. Print swap_avg_by_hour.
3. Use the sorted() function to sort swap_avg_by_hour in descending order. Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.
    * Set the reverse argument to True, so that the highest value in the first column appears first in the list.
    * Assign the result to sorted_swap.
4. Print the string "Top 5 Hours for Ask Posts Comments".
5. Loop through each average and each hour (in this order) in the first five lists of sorted_swap.
6. Use the str.format() method to print the hour and average in the following format: 15:00: 38.59 average comments per post.
    * To format the hours, use the datetime.strptime() constructor to return a datetime object and then use the strftime() method to specify the format of the time.
    * To format the average, you can use {:.2f} to indicate that just two decimal places should be used.
7. Which hours should you create a post during to have a higher chance of receiving comments? Refer back to the documentation for the data set to convert the times to the time zone you live in. Write a markdown cell explaining your findings.


In [16]:
swap_avg_by_hour = []
for lists in avg_by_hour:
    reverse_list =list( reversed(lists))
    swap_avg_by_hour.append(reverse_list)
    
print (swap_avg_by_hour)

[[11.051724137931034, '11'], [7.985294117647059, '23'], [10.08695652173913, '05'], [13.20183486238532, '18'], [14.741176470588234, '13'], [7.170212765957447, '04'], [16.009174311926607, '21'], [23.810344827586206, '02'], [7.796296296296297, '03'], [11.383333333333333, '01'], [5.5777777777777775, '09'], [10.25, '08'], [13.233644859813085, '14'], [8.127272727272727, '00'], [38.5948275862069, '15'], [9.022727272727273, '06'], [13.440677966101696, '10'], [16.796296296296298, '16'], [11.46, '17'], [9.41095890410959, '12'], [7.852941176470588, '07'], [21.525, '20'], [6.746478873239437, '22'], [10.8, '19']]


In [17]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [18]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.


### Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.

That's it for the guided steps! Here's a quick summary of what we accomplished in this guided project:

* We set a goal for the project.
* We collected and sorted the data.
* We reformatted and cleaned the data to prepare it for analysis.
* We analyzed the data.

Guided projects can be used to build a portfolio to showcase to potential employers, so we encourage you to keep working on this. Here are some next steps for you to consider:

* Determine if show or ask posts receive more points on average.
* Determine if posts created at a certain time are more likely to receive more points.
* Compare your results to the average number of comments and points other posts receive.
* Use Dataquest's data science project style guide to format your project.

You're welcome to keep working on the project here, but we recommend downloading it to your computer using the download icon above the notebook and working on it locally.

If you choose to work on the next steps independently, you'll inevitably not know how to perform certain tasks or hit errors that you won't know how to resolve. Don't get discouraged! This is part of the learning process. Although referring back to previous missions is a great way to refresh your memory on certain topics, there are also a couple tools you should get familiar with because you'll need to use them in a real world job setting.

The best thing to do if you hit an error you can't resolve or don't know how to perform a task is search for the answer on Google. When you search, make sure to include the word "python" — otherwise, you'll get results from other programming languages. For example, instead of searching for "how to find the first element in a list," search for "python how to find the first element in a list."

As you search, you'll see one site constantly appear at the top of the results — Stack Overflow. Stack Overflow is an online community where people ask and answer programming questions. In most situations, you'll find that someone has asked the same question as you or a similar question that can help you. The community is very active, so the answers are almost always accurate.

Congratulations, this is the end of the course!